api/py/ai/chronon/repo/validator.py

```diff
@@ -87,52 +103,88 @@ def get_group_by_output_columns(group_by: GroupBy) -> List[str]:
     From the group_by object, get the final output columns after derivations.
     """
     output_columns = set(get_pre_derived_group_by_columns(group_by))
-
     if group_by.derivations:
-        # if derivations contain star, then all columns are included except the columns which are renamed
-        found = any(derivation.expression == "*" for derivation in group_by.derivations)
-        if not found:
-            output_columns.clear()
-        for derivation in group_by.derivations:
-            if found and is_identifier(derivation.expression):
-                output_columns.remove(derivation.expression)
-            if derivation.name != "*":
-                output_columns.add(derivation.name)
-
-    return list(output_columns)
+        return build_derived_columns(output_columns, group_by.derivations)
+    else:
+        return list(output_columns)
 
 
-def get_pre_derived_join_columns(join: Join) -> List[str]:
-    output_columns = []
+def get_pre_derived_join_internal_features(join: Join) -> List[str]:
+    internal_features = []
     for jp in join.joinParts:
-        group_by_cols = get_group_by_output_columns(jp.groupBy)
-        for col in group_by_cols:
+        pre_derived_group_by_features = set(get_pre_derived_group_by_features(jp.groupBy))
+        derived_group_by_features = build_derived_columns(pre_derived_group_by_features, jp.groupBy.derivations)
+        for col in derived_group_by_features:
             prefix = jp.prefix + "_" if jp.prefix else ""
             gb_prefix = jp.groupBy.metaData.name.replace(".", "_")
-            output_columns.append(prefix + gb_prefix + "_" + col)
-    return output_columns
+            internal_features.append(prefix + gb_prefix + "_" + col)
+    return internal_features
 
 
-def get_join_output_columns(join: Join) -> List[str]:
+def get_pre_derived_join_keys(join: Join) -> List[str]:
```

Comment:
this function can be more generalized like 
```
def get_pre_derived_join_keys(source: Source) -> List[str]:
```
Then it can be used [here](https://github.com/airbnb/chronon/pull/777/files#diff-ebaa3fcfb1f15efb4e35e094b3a6dea5d8f01120a3ded8ab53d58fbebf9c0078R86-R91) to avoid duplicate logic

Reply:
Updated